### DISCUSSION TITLE
Generating high frequency FM downsweeps

### DISCUSSION THREAD

**Discussion by raviumadi** [pschatzmann/arduino-audio-tools#1106] [Q&A] [answered]
I am looking for a way to generate downsweeps between 100kHz and 10kHz. This is partly for a project involving echolocating bats/psychophysics. What would be the best way? Is there a generator class? Thanks

**Comment by pschatzmann** [comment, marked as answer]
I am not really sure that you mean with "generate downsweeps". Is this about generating sound or scanning sound in order to display it?

If you want to generate sound, you can simply use one of the sine generators and e.g. change the frequency over time. I think Nyquist-Shannon applies here as well, so you need to use an output sample rate of 20kHz!

**Reply by raviumadi** [reply]
Dear Phil,
Thank you for the reply. I want to generate an FM signal, such as the one attached. Then, I would like to play it at the press of a button via an external DAC (I have PCM5102A). The sampling rate would be at least 192kHz.
Right now, I am testing the following method:
1. Generate a 16-bit, 192kHz wav file in MATLAB with a sweep signal
2. Save to SD
3. Play via ESP32 with example player-sd-i2s.ino modified to replace mp3 decoder with `WAVDecoder`
I also made sample voice recordings at different sampling rates.
The files sampled at 44100 kHz play fine, but at 96kHz, I hear distortion and what sounds like dropouts. 
Perhaps the DAC has a low pass? 

Thanks :-)
![downsweepExample](https://github.com/pschatzmann/arduino-audio-tools/assets/143873837/de47d587-8dcd-48da-9e48-a1947c59fa25)

**Comment by pschatzmann** [comment]
Now you have me complexly confused!
How does your frequency look like over time ? 
What is the singal that you want to modulate with FM ?

**Reply by raviumadi** [reply]
Hi Phil,
Sorry for the confusion.
The same signal is attached as a spectrogram. It is a downward chirp, like a frequency-modulated bat echolocation call.

![downsweepExampleSpec](https://github.com/pschatzmann/arduino-audio-tools/assets/143873837/cd26a07f-f838-4857-bc0f-85c1557d330d)

**Comment by pschatzmann** [comment]
So in this case my initial answer still applies!
This would however for each point of time generate exactly one frequency and not a small band of frequencies.

**Reply by raviumadi** [reply]
Ok. Great! Even though I can configure and play via the I2S port at 192kHz, I still need to learn about the PC5102A DAC. The playback is missing higher frequencies. When I play white noise generated at 192kHz, a live spectrogram shows a cutoff at ~20kHz. Do you have any insights about the DAC? The TI documentation says the chip is capable of fs up to 384kHz.
Thanks :-)

**Comment by pschatzmann** [comment]
That information should be available in the technical sheet.
I guess it will be pretty difficult to find some speakers that cover that range...
Did you check the frequency range of you measuring equipment ?

**Reply by raviumadi** [reply]
I see. Yeah, I have the proper equipment for this. I also have speakers calibrated up to 96kHz for flat frequency response against a 1/4" B&K reference microphone.

**Comment by pschatzmann** [comment]
I have seen that it is possible to generate C code from Matlab. 
So you could use that a basis to write your own sound generator subclass

**Reply by raviumadi** [reply]
That is a solid idea. Thank you :-)
